### ISSUE TITLE
kenel panic in MergingRavennaALSA module when ALSA device is used by speaker_test at 96KHZ

### ISSUE THREAD

**Issue by siraaris** [bondagit/aes67-linux-daemon#122]
I've got a reproducible kernel panic happening and don't really know where/how to debug further. 

My setup is a central DSP server with 4 endpoints. 

The server handles 32 ch, and each of the endpoints 8 ch each. 

Panic happens on Linux 6.1 and 5.19. 

It occurs immediately on the server once the DSP process writes sound to RAVENNA PCM device.

Any help or hints how to debug or investigate would be much appreciated. 

Aris

**The following occurs when a DSP process (CamillaDSP) attempts to send 32ch audio to the RAVENNA ALSA device:**

![image](https://user-images.githubusercontent.com/12581268/227396081-d2b2ce15-543c-4e56-a1bf-71aa3d34f225.jpeg)

**Comment by siraaris** [comment]
A bit more information - dmesg when a panic didn't happen immediately:

[30210.407665] entering mr_alsa_audio_pcm_open (substream name=subdevice #0 #0) ...
[30210.407775] 
               Current PTPFrame Size = 96
[30210.407776] minPTPFrameSize = 48
[30210.407777] maxPTPFrameSize = 384
[30210.407853] mr_alsa_audio_pcm_hw_params (enter): rate=96000 format=32 channels=32 period_size=96, nb_periods=1024
               , buffer_bytes=9437184
[30210.407855] mr_alsa_audio_pcm_hw_params (playback): wrong nbPeriods (1024 instead of 512)...
[30210.408520] mr_alsa_audio_pcm_hw_params done: rate=96000 format=32 channels=32 period_size=96, nb_periods=1024
               , buffer_bytes=9437184
[30210.408847] entering mr_alsa_audio_pcm_prepare (substream name=subdevice #0 #0) ...
[30210.408848] mr_alsa_audio_pcm_prepare: rate=96000 format=32 channels=32 period_size=96, nb periods=1024
[30210.408850] mr_alsa_audio_pcm_prepare for playback stream
[30210.408850] Leaving mr_alsa_audio_pcm_prepare..
[30210.417172] entering mr_alsa_audio_pcm_trigger (substream name=subdevice #0 #0) ...
[30210.417174] mr_alsa_audio_pcm_trigger(Start), rate=96000 format=32 channels=32 period_size=96
[30210.417175] starting playback I/O
[30210.418471] BUG: unable to handle page fault for address: ffffa1fec3b5c000
[30210.418474] #PF: supervisor write access in kernel mode
[30210.418475] #PF: error_code(0x0002) - not-present page
[30210.418477] PGD 100000067 P4D 100000067 PUD 1001a5067 PMD 1381d0067 PTE 0
[30210.418480] Oops: 0002 [#1] PREEMPT_RT SMP NOPTI
[30210.418482] CPU: 15 PID: 9623 Comm: speaker-test Tainted: G           OE     5.19.11-rt9-cexe #1 
[30210.418484] Hardware name: Supermicro Super Server/X12SCV-W, BIOS 2.0 07/14/2021
[30210.418485] RIP: 0010:memset_erms+0x9/0x20
[30210.418490] Code: 01 48 0f af c6 f3 48 ab 89 d1 f3 aa 4c 89 c8 c3 cc cc cc cc 66 66 2e 0f 1f 84 00 00 00 00 00 66 90 49 89 f9 40 88 f0 48 89 d1 <f3> aa 4c 89 c8 c3 cc cc cc cc 66 66 2e 0f 1f 84 00 00 00 00 00 66
[30210.418491] RSP: 0018:ffffa1fec6023ca0 EFLAGS: 00010287
[30210.418493] RAX: 0000000000000000 RBX: ffffffffc0a19ca0 RCX: 0000000001800000
[30210.418494] RDX: 0000000003000000 RSI: 0000000000000000 RDI: ffffa1fec3b5c000
[30210.418495] RBP: 0000000000018000 R08: ffffffffb026bde0 R09: ffffa1fec235c000
[30210.418496] R10: 63616279616c7020 R11: 676e697472617473 R12: ffff96400120da18
[30210.418497] R13: ffff963f87addc00 R14: 0000000000000060 R15: ffff963f86640000
[30210.418498] FS:  00007fd9fd6b4b80(0000) GS:ffff964e9c5c0000(0000) knlGS:0000000000000000
[30210.418499] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[30210.418500] CR2: ffffa1fec3b5c000 CR3: 0000000134b9a005 CR4: 00000000007706e0
[30210.418501] PKRU: 55555554
[30210.418502] Call Trace:
[30210.418503]  <TASK>
[30210.418505]  MuteOutputBuffer+0xa8/0xf0 [MergingRavennaALSA]
[30210.418513]  start_interrupts+0x49/0x67 [MergingRavennaALSA]
[30210.418518]  mr_alsa_audio_pcm_trigger+0x13a/0x183 [MergingRavennaALSA]
[30210.418523]  ? rt_spin_unlock+0x13/0x40
[30210.418527]  ? find_vmap_area+0x3d/0x60
[30210.418532]  snd_pcm_action_single+0x33/0x80 [snd_pcm]
[30210.418539]  __snd_pcm_lib_xfer+0x805/0xa30 [snd_pcm]
[30210.418547]  ? snd_pcm_hw_rule_step+0x100/0x100 [snd_pcm]
[30210.418553]  ? snd_pcm_hw_constraint_mask64+0x60/0x60 [snd_pcm]
[30210.418560]  snd_pcm_common_ioctl+0xc2f/0x1360 [snd_pcm]
[30210.418567]  snd_pcm_ioctl+0x23/0x40 [snd_pcm]
[30210.418573]  __x64_sys_ioctl+0x8d/0xd0
[30210.418576]  do_syscall_64+0x37/0xc0
[30210.418579]  entry_SYSCALL_64_after_hwframe+0x63/0xcd
[30210.418582] RIP: 0033:0x7fd9fd7b3bab
[30210.418584] Code: 00 48 89 44 24 18 31 c0 48 8d 44 24 60 c7 04 24 10 00 00 00 48 89 44 24 08 48 8d 44 24 20 48 89 44 24 10 b8 10 00 00 00 0f 05 <89> c2 3d 00 f0 ff ff 77 1c 48 8b 44 24 18 64 48 2b 04 25 28 00 00
[30210.418585] RSP: 002b:00007ffc2a6054b0 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[30210.418587] RAX: ffffffffffffffda RBX: 000055c604644da0 RCX: 00007fd9fd7b3bab
[30210.418588] RDX: 00007ffc2a605510 RSI: 0000000040184150 RDI: 0000000000000004
[30210.418589] RBP: 000055c604644d20 R08: 000055c602f150c0 R09: 0000000000000073
[30210.418590] R10: 0000000000000000 R11: 0000000000000246 R12: 000055c60464b160
[30210.418590] R13: 000055c604644da0 R14: 0000000000000060 R15: 0000000000000bb8
[30210.418592]  </TASK>
[30210.418593] Modules linked in: veth xt_nat xt_tcpudp xt_conntrack nft_chain_nat xt_MASQUERADE nf_nat nf_conntrack_netlink nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 xfrm_user xfrm_algo xt_addrtype nft_compat nf_tables libcrc32c nfnetlink br_netfilter bridge stp llc msr overlay intel_rapl_msr binfmt_misc intel_rapl_common nls_ascii nls_cp437 vfat fat x86_pkg_temp_thermal snd_sof_pci_intel_cnl intel_powerclamp coretemp snd_sof_intel_hda_common soundwire_intel soundwire_generic_allocation soundwire_cadence snd_sof_intel_hda snd_sof_pci kvm_intel snd_sof_xtensa_dsp snd_sof snd_sof_utils i915 snd_soc_hdac_hda snd_hda_ext_core snd_soc_acpi_intel_match snd_soc_acpi drm_buddy kvm snd_soc_core drm_display_helper snd_compress cec irqbypass snd_intel_dspcfg snd_intel_sdw_acpi ghash_clmulni_intel rc_core snd_hda_codec mei_wdt aesni_intel ttm mei_hdcp crypto_simd cryptd snd_hda_core rapl mei_me intel_cstate iTCO_wdt snd_hwdep intel_pmc_bxt iTCO_vendor_support intel_uncore soundwire_bus
[30210.418625]  drm_kms_helper wmi_bmof watchdog pcspkr ee1004 mei intel_pch_thermal evdev intel_pmc_core acpi_tad acpi_pad button apple_mfi_fastcharge sg nfsd auth_rpcgss snd_aloop nfs_acl lockd MergingRavennaALSA(OE) grace drm snd_pcm sunrpc snd_timer snd fuse soundcore loop dm_mod efi_pstore configfs ip_tables x_tables autofs4 ext4 crc16 mbcache jbd2 crc32c_generic hid_apple hid_generic usbhid hid sd_mod ahci libahci xhci_pci nvme libata xhci_hcd nvme_core t10_pi igb scsi_mod usbcore crc64_rocksoft igc crc64 crc32_pclmul i2c_i801 crc_t10dif crc32c_intel crct10dif_generic i2c_smbus i2c_algo_bit crct10dif_pclmul scsi_common crct10dif_common usb_common dca fan wmi video
[30210.418660] CR2: ffffa1fec3b5c000
[30210.430800] ---[ end trace 0000000000000000 ]---
[30210.478399] RIP: 0010:memset_erms+0x9/0x20
[30210.478403] Code: 01 48 0f af c6 f3 48 ab 89 d1 f3 aa 4c 89 c8 c3 cc cc cc cc 66 66 2e 0f 1f 84 00 00 00 00 00 66 90 49 89 f9 40 88 f0 48 89 d1 <f3> aa 4c 89 c8 c3 cc cc cc cc 66 66 2e 0f 1f 84 00 00 00 00 00 66
[30210.478405] RSP: 0018:ffffa1fec6023ca0 EFLAGS: 00010287
[30210.478408] RAX: 0000000000000000 RBX: ffffffffc0a19ca0 RCX: 0000000001800000
[30210.478409] RDX: 0000000003000000 RSI: 0000000000000000 RDI: ffffa1fec3b5c000
[30210.478411] RBP: 0000000000018000 R08: ffffffffb026bde0 R09: ffffa1fec235c000
[30210.478411] R10: 63616279616c7020 R11: 676e697472617473 R12: ffff96400120da18
[30210.478412] R13: ffff963f87addc00 R14: 0000000000000060 R15: ffff963f86640000
[30210.478413] FS:  00007fd9fd6b4b80(0000) GS:ffff964e9c5c0000(0000) knlGS:0000000000000000
[30210.478415] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[30210.478416] CR2: ffffa1fec3b5c000 CR3: 0000000134b9a005 CR4: 00000000007706e0
[30210.478417] PKRU: 55555554
[30231.420730] rcu: INFO: rcu_preempt detected stalls on CPUs/tasks:
[30231.420732] rcu: 	Tasks blocked on level-0 rcu_node (CPUs 0-15): P9623/2:b..l
[30231.420736] 	(detected by 0, t=5252 jiffies, g=604621, q=3240 ncpus=16)
[30231.420738] task:speaker-test    state:D stack:    0 pid: 9623 ppid:  9611 flags:0x00004002
[30231.420741] Call Trace:
[30231.420742]  <TASK>
[30231.420744]  __schedule+0x359/0xa30
[30231.420749]  schedule_rtlock+0x1e/0x40
[30231.420751]  rtlock_slowlock_locked+0x3c3/0xf80
[30231.420755]  ? rt_spin_unlock+0x13/0x40
[30231.420758]  rt_spin_lock+0x3f/0x60
[30231.420761]  snd_pcm_drop+0x4f/0xe0 [snd_pcm]
[30231.420770]  snd_pcm_release_substream.part.0+0xf/0xd0 [snd_pcm]
[30231.420775]  snd_pcm_release+0x5a/0xc0 [snd_pcm]
[30231.420781]  __fput+0x87/0x240
[30231.420784]  task_work_run+0x59/0x90
[30231.420788]  do_exit+0x323/0xb10
[30231.420790]  make_task_dead+0x60/0x60
[30231.420791]  rewind_stack_and_make_dead+0x17/0x20
[30231.420794] RIP: 0033:0x7fd9fd7b3bab
[30231.420796] RSP: 002b:00007ffc2a6054b0 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[30231.420798] RAX: ffffffffffffffda RBX: 000055c604644da0 RCX: 00007fd9fd7b3bab
[30231.420799] RDX: 00007ffc2a605510 RSI: 0000000040184150 RDI: 0000000000000004
[30231.420800] RBP: 000055c604644d20 R08: 000055c602f150c0 R09: 0000000000000073
[30231.420801] R10: 0000000000000000 R11: 0000000000000246 R12: 000055c60464b160
[30231.420802] R13: 000055c604644da0 R14: 0000000000000060 R15: 0000000000000bb8
[30231.420804]  </TASK>

**Comment by siraaris** [comment]
I have reproduced that a panic occurs on a non-RT kernel as well.

**Comment by bondagit** [comment]
can you please try to use the driver in branch [aes67-daemon-issue122](https://github.com/bondagit/ravenna-alsa-lkm/tree/aes67-daemon-issue122) of the [ravenna-alsa-lkm](https://github.com/bondagit/ravenna-alsa-lkm) repo.

**Comment by siraaris** [comment]
The above branch still panics.
![IMG_0971](https://user-images.githubusercontent.com/12581268/227613741-c89f8a83-2c5b-44b2-b18d-ab376e7c7ba7.jpeg)

**Comment by bondagit** [comment]
I can reproduce the issue and It looks like there is a problem when the _speaker-test_ application is used at 96Khz.

Return to the original driver version (in branch aes67-daemon) and try to run the platform and latency tests:

`./run_test.sh S24_3LE 96000 8 1`

and

`./run_latency_test.sh S24_3LE 96000 8 10 128`

If these work properly for you try to use _aplay_ instead of speak-test:

`aplay -D plughw:RAVENNA -f S24_3LE -r 96000 -c 8 -vv test.wav`
 
All the above tests work properly for me.

**Comment by siraaris** [comment]
First step:

`root@controller:~/proj/aes67-linux-daemon# ./run_latency_test.sh S24_3LE 96000 8 10 128
Compiling tools ...
g++    -c -o check.o check.cc
g++   check.o   -o check
g++    -c -o createtest.o createtest.cc
g++   createtest.o   -o createtest
g++    -c -o latency.o latency.cc
g++ latency.o -o latency  -lpthread -lasound
Using buffer size of 128 frames
Creating configuration files ..
net.ipv4.igmp_max_memberships = 66
kernel.perf_cpu_time_max_percent = 0
kernel.sched_rt_runtime_us = 1000000
Starting PTP master ...
Starting AES67 daemon ...
Waiting for PTP slave to sync ...
Running 10 secs latency test
Scheduler set to Round Robin with priority 99...
Playback device is hw:RAVENNA
Capture device is hw:RAVENNA
Parameters are 96000Hz, S24_3LE, 8 channels, non-blocking mode
Poll mode: no
Loop limit is 960000 frames, minimum latency = 128, maximum latency = 128
Hardware PCM card 0 'Merging RAVENNA' device 0 subdevice 0
Its setup is:
  stream       : PLAYBACK
  access       : RW_INTERLEAVED
  format       : S24_3LE
  subformat    : STD
  channels     : 8
  rate         : 96000
  exact rate   : 96000 (96000/1)
  msbits       : 24
  buffer_size  : 256
  period_size  : 128
  period_time  : 1333
  tstamp_mode  : NONE
  tstamp_type  : MONOTONIC
  period_step  : 1
  avail_min    : 128
  period_event : 0
  start_threshold  : 2147483647
  stop_threshold   : 256
  silence_threshold: 0
  silence_size : 0
  boundary     : 4611686018427387904
  appl_ptr     : 0
  hw_ptr       : 0
Hardware PCM card 0 'Merging RAVENNA' device 0 subdevice 0
Its setup is:
  stream       : CAPTURE
  access       : RW_INTERLEAVED
  format       : S24_3LE
  subformat    : STD
  channels     : 8
  rate         : 96000
  exact rate   : 96000 (96000/1)
  msbits       : 24
  buffer_size  : 256
  period_size  : 128
  period_time  : 1333
  tstamp_mode  : NONE
  tstamp_type  : MONOTONIC
  period_step  : 1
  avail_min    : 128
  period_event : 0
  start_threshold  : 2147483647
  stop_threshold   : 256
  silence_threshold: 0
  silence_size : 0
  boundary     : 4611686018427387904
  appl_ptr     : 0
  hw_ptr       : 0
Trying latency 256 frames, 2666.667us, 2.666667ms (375.0000Hz)
Success
Playback:
*** frames = 960256 ***
  state       : RUNNING
  trigger_time: 5531.415063
  tstamp      : 0.000000
  delay       : 256
  avail       : 0
  avail_max   : 128
Capture:
*** frames = 960000 ***
  state       : RUNNING
  trigger_time: 5531.415064
  tstamp      : 0.000000
  delay       : 0
  avail       : 0
  avail_max   : 128
Maximum read: 128 frames
Maximum read latency: 1333.333us, 1.333333ms (750.0000Hz)
Playback time = 5531.415063, Record time = 5531.415064, diff = -1
End to end latency: 14.666 msecs
Terminating processes ...
daemon exiting with code: 0
`

**Comment by siraaris** [comment]
aplay does indeed seem to work. 

On that basis, let's close this please. 

Thank you for your investigation. 

Kind regards,

Aris

**Comment by bondagit** [comment]
I fixed the issue in the driver and speaker-test should now work properly at 96Khz and above. 
Can you please checkout the daemon branch [issue-124](https://github.com/bondagit/aes67-linux-daemon/issues/114) and retry ? 
Please consider you also need to use the new driver branch [aes67-daemon-issue11](https://github.com/bondagit/ravenna-alsa-lkm/tree/aes67-daemon-issue11)

**Comment by siraaris** [comment]
The ./run_* scripts for 96000 and 192000 / 8ch run fine.

OK  ./run_test.sh S24_3LE 96000 8 1
OK  ./run_latency_test.sh S24_3LE 96000 8 10 128

and

OK  ./run_test.sh S24_3LE 192000 8 1
OK  ./run_latency_test.sh S24_3LE 192000 8 10 128

Increasing to 32 ch results in overrun and error.

root@controller:~/proj/aes67-linux-daemon# ./run_test.sh S24_3LE 192000 32 1
Compiling tools ...
make: Nothing to be done for 'all'.
Creating test file ...
test file created
Creating configuration files ..
net.ipv4.igmp_max_memberships = 66
kernel.perf_cpu_time_max_percent = 0
kernel.sched_rt_runtime_us = 1000000
Starting PTP master ...
Starting AES67 daemon ...
Waiting for PTP slave to sync ...
Starting to record 60 sec from sink ...
Recording raw data '/tmp/sink_test.raw' : Signed 24 bit Little Endian in 3bytes, Rate 192000 Hz, Channels 32
Starting to playback test on source ...
overrun!!! (at least 484.128 ms long)
overrun!!! (at least 24.734 ms long)
overrun!!! (at least 559.882 ms long)
overrun!!! (at least 52.523 ms long)
overrun!!! (at least 12.436 ms long)
overrun!!! (at least 116.431 ms long)
Recording to file "test/sink_test.raw" successfull
Test result:
error at position: 465647617
Terminating processes ...
daemon exiting with code: 0

I tried that as in my project I'm running 32ch @ 192khz, which then split out to 2 x 8ch @192khz.

Trying speaker-test @ 96Khz:

root@controller:~/proj/aes67-linux-daemon# speaker-test -D hw:RAVENNA -r 96000 -c 8 -f 1000 -F S24_3LE

speaker-test 1.2.8

Playback device is hw:RAVENNA
Stream parameters are 96000Hz, S24_3LE, 8 channels
Using 16 octaves of pink noise
Rate set to 96000Hz (requested 96000Hz)
Buffer size range from 192 to 12288
Period size range from 96 to 96
Using max buffer size 12288
Periods = 4
Unable to set hw params for playback: Invalid argument
Setting of hwparams failed: Invalid argument

root@controller:~/proj/aes67-linux-daemon# speaker-test -D hw:RAVENNA -r 192000 -c 8 -f 1000 -F S24_3LE

speaker-test 1.2.8

Playback device is hw:RAVENNA
Stream parameters are 192000Hz, S24_3LE, 8 channels
Using 16 octaves of pink noise
Rate set to 192000Hz (requested 192000Hz)
Buffer size range from 384 to 24576
Period size range from 192 to 192
Using max buffer size 24576
Periods = 4
was set period_size = 192
was set buffer_size = 24576
 0 - Front Left
Write error: -5,Input/output error
xrun_recovery failed: -5,Input/output error
Transfer failed: Input/output error


with dmesg:

[526866.014370] entering mr_alsa_audio_pcm_open (substream name=subdevice #0 #0) ...
[526866.014373] mr_alsa_audio_pcm_open: playback format nb bits range: [8, 32]
[526866.014374] mr_alsa_audio_pcm_open: playback period size range: [48, 196608], periods range: [2, 128]
[526866.014375] mr_alsa_audio_pcm_open: get_master_switch_value error
[526866.014376] Current PTPFrame Size = 192, minPTPFrameSize = 48, maxPTPFrameSize = 384
[526866.014449] mr_alsa_audio_pcm_hw_params (enter): rate=192000 format=32 channels=8 period_size=192, nb_periods=128
                , buffer_bytes=589824
[526866.014451] mr_alsa_audio_pcm_hw_params (playback): wrong nbPeriods (128 instead of 256)...
[526866.014494] mr_alsa_audio_pcm_hw_params done: rate=192000 format=32 channels=8 period_size=192, nb_periods=128
                , buffer_bytes=589824
[526866.014519] entering mr_alsa_audio_pcm_prepare (substream name=subdevice #0 #0) ...
[526866.014520] mr_alsa_audio_pcm_prepare: rate=192000 format=32 channels=8 period_size=192, nb periods=128
[526866.014521] mr_alsa_audio_pcm_prepare for playback stream
[526866.014522] Leaving mr_alsa_audio_pcm_prepare..
[526866.015204] entering mr_alsa_audio_pcm_trigger (substream name=subdevice #0 #0) ...
[526866.015205] mr_alsa_audio_pcm_trigger(Start), rate=192000 format=32 channels=8 period_size=192
[526876.021646] entering mr_alsa_audio_pcm_trigger (substream name=subdevice #0 #0) ...
[526876.021648] mr_alsa_audio_pcm_trigger(Stop), rate=192000 format=32 channels=8 period_size=192
[526876.021652] entering mr_alsa_audio_pcm_hw_free (substream name=subdevice #0 #0) ...
[526876.021694] entering mr_alsa_audio_pcm_close (substream name=subdevice #0 #0) ...

**Comment by bondagit** [comment]
If the platform test fails with 32 channels at 192khz it probably means you cannot safely use this configuration. Try to see if it works with 16 channels. The platform test runs on the loopback network device, it starts the daemon and a master clock automatically. 
When you use speaker-test make sure you start the daemon and the PTP slave status turns to locked otherwise you are not able to payback or capture on the Ravenna device because the audio device is not clocked and you have the transfer error you see To achieve this you need to have a master clock on the network. You can use the daemon WebUI to check for the PTP slave status.

**Comment by siraaris** [comment]
For me, 96k x 8ch works, but not 16 or 32.

BTW, the text for branch mentions "issue-124" but links to 114.

**Comment by bondagit** [comment]
Yes, thanks for the note I fixed the link name.

**Comment by siraaris** [comment]
So I'm using an i7, SSD with 64gb RAM on a dev system here and can't get anything above 16x48khz S24_3SE.

Something is weird.

**Comment by bondagit** [comment]
The ticket is about being able to run speaker-test application @96Khz and higher sample rates without a kernel hang.
Your system max channels investigation is a different topic. Does speaker-test with 2 channels work for you now?

**Comment by siraaris** [comment]
Confirming that "speaker-test -D hw:RAVENNA -r 96000 -c 16 -f 1000 -F S24_3LE"  works on the above branches.

Thank you!

**Comment by bondagit** [comment]
> So I'm using an i7, SSD with 64gb RAM on a dev system here and can't get anything above 16x48khz S24_3SE.
> 
> Something is weird.

yes, you are right and I can reproduce this issue. The patch applied to the driver in the above branch is causing the platform test to fail  with anything above 16x48khz S24_3SE.

**Comment by siraaris** [comment]
That's good to know. I was worried that test failure would be an indication of a system issue, but that doesn't seem to affect actual performance, I'm running 4 x 8 streams @96KHz and it's quite stable and performant.

**Comment by bondagit** [comment]
I have found the issue causing the problem and I have just released a new patch for the driver.
Can you please checkout the daemon branch [issue-114](https://github.com/bondagit/aes67-linux-daemon/tree/issue-114) and retry ?
Please consider you also need to use the new driver branch [aes67-daemon-issue11](https://github.com/bondagit/ravenna-alsa-lkm/tree/aes67-daemon-issue11) .
The build.sh script in the new daemon branch will checkout the correct driver.

**Comment by bondagit** [comment]
solved starting from driver v1.7

**Comment by siraaris** [comment]
Apologies if commenting will re-open this one - just confirming that 1.7 is yet to be released - current is at 1.6.5 correct?

**Comment by siraaris** [comment]
Ok, understand that 1.7 is the LKM driver. I'm just investigating a crasher (kernel panic) at 192000 on most recent versions of aes67-daemon and driver - will look into further and raise another issue if required.
